utils/swift_build_support/swift_build_support/targets.py

```diff
@@ -12,60 +12,124 @@
 import platform
 
 
-class StdlibDeploymentTarget(object):
+class Platform(object):
+    """
+    Abstract representation of a platform Swift can run on.
+    """
 
-    class OSX(object):
-        x86_64 = 'macosx-x86_64'
-        allArchs = [x86_64]
-
-    class iOS(object):  # noqa
-        armv7 = 'iphoneos-armv7'
-        armv7s = 'iphoneos-armv7s'
-        arm64 = 'iphoneos-arm64'
-        allArchs = [armv7, armv7s, arm64]
-
-    class iOSSimulator(object):  # noqa
-        i386 = 'iphonesimulator-i386'
-        x86_64 = 'iphonesimulator-x86_64'
-        allArchs = [i386, x86_64]
-
-    class AppleTV(object):
-        arm64 = 'appletvos-arm64'
-        allArchs = [arm64]
-
-    class AppleTVSimulator(object):
-        x86_64 = 'appletvsimulator-x86_64'
-        allArchs = [x86_64]
-
-    class AppleWatch(object):
-        armv7k = 'watchos-armv7k'
-        allArchs = [armv7k]
-
-    class AppleWatchSimulator(object):
-        i386 = 'watchsimulator-i386'
-        allArchs = [i386]
-
-    class Linux(object):
-        x86_64 = 'linux-x86_64'
-        armv6 = 'linux-armv6'
-        armv7 = 'linux-armv7'
-        aarch64 = 'linux-aarch64'
-        ppc64 = 'linux-ppc64'
-        ppc64le = 'linux-ppc64le'
-        s390x = 'linux-s390x'
-        allArchs = [x86_64, armv6, armv7, aarch64, ppc64, ppc64le, s390x]
-
-    class FreeBSD(object):
-        amd64 = 'freebsd-x86_64'
-        allArchs = [amd64]
-
-    class Cygwin(object):
-        x86_64 = 'cygwin-x86_64'
-        allArchs = [x86_64]
-
-    class Android(object):
-        armv7 = 'android-armv7'
-        allArchs = [armv7]
+    def __init__(self, name, archs, sdk_name=None):
+        """
+        Create a platform with the given name and list of architectures.
+        """
+        self.name = name
+        self.targets = [Target(self, arch) for arch in archs]
+        # FIXME: Eliminate this argument; apparently the SDK names are
+        # internally a private implementation detail of the build script, so we
+        # should just make them the same as the platform name.
+        self.sdk_name = name.upper() if sdk_name is None else sdk_name
+
+        # Add a property for each arch.
+        for target in self.targets:
+            setattr(self, target.arch, target)
+
+    @property
+    def is_darwin(self):
+        """Convenience function for checking if this is a Darwin platform."""
+        return isinstance(self, DarwinPlatform)
+
```

Comment:
I don't feel like this conversation is continuing very productively, as I have said I agree with your direction but I don't personally have time to implement all refactors that I agree are useful. IMHO, code duplication is one of the most painful things to refactor because you have to verify the new common code behavior versus all the divergent paths, and that is what I did here.

I'll leave it up to @gribozavr whether he wants this to be refined before landing.